src/trustline_asset.js

```diff
@@ -0,0 +1,206 @@
+import xdr from './generated/stellar-xdr_generated';
+import { Keypair } from './keypair';
+import { StrKey } from './strkey';
+
+/**
+ * TrustLineAsset class represents a trustline to either a liquidity pool, the
+ * native asset (`XLM`) or an issued asset with an asset code / issuer account
+ * ID pair.
+ *
+ * In case of the native asset, the code will represent `XLM` while the issuer
+ * and liquidityPoolId will be empty. For an issued asset, the code and issuer
+ * will be valid and liquidityPoolId will be empty. For liquidity pools, the
+ * liquidityPoolId will be valid and the remaining two fields will be empty.
+ *
+ * @constructor
+ * @param {string} code - The asset code.
+ * @param {string} issuer - The account ID of the asset issuer.
+ * @param {string} liquidityPoolId - The ID of the liquidity pool in string 'hex'.
+ */
+export class TrustLineAsset {
+  constructor(code, issuer, liquidityPoolId) {
+    if (!code && !issuer && !liquidityPoolId) {
+      throw new Error('Must provide either code, issuer or liquidityPoolId');
+    }
+    if ((code || issuer) && !/^[a-zA-Z0-9]{1,12}$/.test(code)) {
+      throw new Error(
+        'Asset code is invalid (maximum alphanumeric, 12 characters at max)'
+      );
+    }
+    if (code && String(code).toLowerCase() !== 'xlm' && !issuer) {
+      throw new Error('Issuer cannot be null');
+    }
+    if (issuer && !StrKey.isValidEd25519PublicKey(issuer)) {
+      throw new Error('Issuer is invalid');
+    }
+
+    // Test if liquiditry pool ID is a hash.
+    if (liquidityPoolId && !/^[a-fA-F0-9]{64}$/.test(liquidityPoolId)) {
+      throw new Error('Liquidity pool ID is not a valid hash');
+    }
+
+    this.code = code;
+    this.issuer = issuer;
+    this.liquidityPoolId = liquidityPoolId;
+  }
+
+  /**
+   * Returns a trustline asset object for the native asset.
+   * @returns {TrustLineAsset}
+   */
+  static native() {
+    return new TrustLineAsset('XLM');
+  }
+
+  /**
+   * Returns a trustline asset object from its XDR object representation.
+   * @param {xdr.TrustLineAsset} tlAssetXdr - The asset XDR object.
+   * @returns {TrustLineAsset}
+   */
+  static fromOperation(tlAssetXdr) {
+    let anum;
+    let code;
+    let issuer;
+    let liquidityPoolId;
+    switch (tlAssetXdr.switch()) {
+      case xdr.AssetType.assetTypeNative():
+        return this.native();
+      case xdr.AssetType.assetTypeCreditAlphanum4():
+        anum = tlAssetXdr.alphaNum4();
+      /* falls through */
+      case xdr.AssetType.assetTypeCreditAlphanum12():
+        anum = anum || tlAssetXdr.alphaNum12();
+        issuer = StrKey.encodeEd25519PublicKey(anum.issuer().ed25519());
+        code = trimEnd(anum.assetCode(), '\0');
+        return new this(code, issuer);
+      case xdr.AssetType.assetTypePoolShare():
+        // TODO: review if this is correct
+        liquidityPoolId = tlAssetXdr.liquidityPoolId().toString('hex');
+        return new this(null, null, liquidityPoolId);
```

Comment:
👀

Reply:
Not sure if we are doing hex here?

Reply:
That was a tricky one for me too and I'm not sure this is correct... I did it based on the transaction ID whose representation is a hash-hex of the transaction.

How do you represent liquidity pool IDs in Core @jonjove?

Reply:
We represent it as the raw bytes (https://datatracker.ietf.org/doc/html/rfc4506.html#section-4.9) of the hash. It should be represented the same as `ClaimableBalanceID.v0()`. I have basically no context on this code base so I can't easily comment on whether this is right or wrong. I don't know if it will be helpful, but we have tests for liquidity pool IDs in https://github.com/stellar/stellar-core/pull/3165/commits/5e8d7cfe81cc79978a43d39f075f719dae5e2a09.

Reply:
Sharing those samples will be very helpful 🙌 , thank you!

IDs like `c17f36fbd210e43dca1cda8edc5b6c0f825fcb72b39f0392fd6309844d77ff7d` do look like a hex representation but I'll double-check it later. Also, I'll make sure to add tests with the same assets and expected poolID to make sure the JS SDK is using the same logic as core.

Reply:
The tests convert to hex for human readability and writability (it's hard to write binary strings directly into a program). You should convert the hex strings back to raw bytes.

Reply:
Hey @jonjove, I've added a few tests in https://github.com/stellar/js-stellar-base/pull/454/commits/54e0070d772fc02745096530c06ddd998860123c mirroring the same tests from stellar-core. We're reaching the same results so I'm confident that the way we're calculating the liquidity pool ID is the same 🙌 .

About hex vs raw bytes, I'm handling the pool ID the same way we already handle claimable balance IDs so I believe we're fine there. The PoolID is represented as a 'hex' string to make it more readable but we convert it to raw bytes when converting the objects to XDR.